src/observers/strategies/cache_occupancy.cpp

```diff
@@ -0,0 +1,81 @@
+#include <list>
+#include <utility>
+#include <vector>
+
+#include "bus/event_bus.hpp"
+
+#include "observers/strategies/seniority.hpp"
+
+#include "serenity/resource_helper.hpp"
+
+#include "cache_occupancy.hpp"
+
+namespace mesos {
+namespace serenity {
+
+using std::list;
+using std::pair;
+
+Try<QoSCorrections> CacheOccupancyStrategy::decide(
+    ExecutorAgeFilter* ageFilter,
+    const Contentions& contentions,
+    const ResourceUsage& usage) {
+  std::vector<ResourceUsage_Executor> cmtEnabledExecutors =
+      getCmtEnabledExecutors(usage);
+
+  if (cmtEnabledExecutors.empty()) {
+    return QoSCorrections();
+  }
+
+  double_t meanCacheOccupancy = countMeanCacheOccupancy(cmtEnabledExecutors);
+  std::vector<ResourceUsage_Executor> aggressors =
+    getExecutorsAboveMeanCacheOccupancy(cmtEnabledExecutors,
+                                        meanCacheOccupancy);
+
+  QoSCorrections corrections;
+  for (auto aggressor : aggressors) {
+    ExecutorInfo executorInfo = aggressor.executor_info();
+    corrections.push_back(createKillQosCorrection(executorInfo));
+  }
+
+  return corrections;
+}
+
+std::vector<ResourceUsage_Executor>
+CacheOccupancyStrategy::getCmtEnabledExecutors(
+    const ResourceUsage& usage) const {
+  std::vector<ResourceUsage_Executor> executors;
+  for (const ResourceUsage_Executor& executor : usage.executors()) {
+    if (executor.has_statistics() &&
+        executor.statistics().has_perf() &&
+        executor.statistics().perf().has_llc_occupancy()) {
```

Comment:
Unfortunately we need either define here or upstream ResourceUsage with llc_occupancy in Mesos.. Otherwise users would have to use our `patched` mesos (: